File: app/src/main/java/com/example/thegoldenratio/facerecognition/faceRecognition.java
package com.example.thegoldenratio.facerecognition;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.graphics.drawable.BitmapDrawable;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.provider.MediaStore;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.core.content.FileProvider;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProvider;

import com.example.thegoldenratio.R;
import com.example.thegoldenratio.databinding.ActivityFaceRecognitionBinding;

import java.io.File;
import java.io.IOException;

public class faceRecognition extends AppCompatActivity {


    private static final float STROKE_WIDTH = 6.0f;

    private FaceRecViewModel viewModel;
    private ActivityFaceRecognitionBinding binding;

    public int index = 1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        getPermissions();

        binding = DataBindingUtil.setContentView(this, R.layout.activity_face_recognition);
        viewModel = new ViewModelProvider(this).get(FaceRecViewModel.class);

        binding.faceRecCameraFloatButton.setOnClickListener(v -> launchCameraIntent());
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent intent) {
        super.onActivityResult(requestCode, resultCode, intent);

        if (resultCode == RESULT_OK && requestCode == 1) {
            final Bitmap bitmap = BitmapFactory.decodeFile(viewModel.currentPhotoPath);
            if (bitmap != null) {
                binding.imageView.setVisibility(View.VISIBLE);
                viewModel.mSelectedImage = Bitmap.createScaledBitmap(bitmap, binding.imageView.getWidth(), binding.imageView.getHeight(), true);
                binding.imageView.setImageBitmap(viewModel.mSelectedImage);

                viewModel.analyzeImage();

                try {
                    viewModel.analyzeJson(viewModel.resultString);
                    binding.forwardArrow.setOnClickListener(v -> {
                        index++;
                        if(index > 5) index = 1;
                        changeRectangle(index);
                    });
                    binding.backWardArrow.setOnClickListener(v -> {
                        index--;
                        if(index < 1) index = 5;
                        changeRectangle(index);
                    });
                }catch (Exception e){
                    Log.i("result", e.toString());
                }
            }
        }
    }

    public void getPermissions(){
        if(ContextCompat.checkSelfPermission(faceRecognition.this, Manifest.permission.CAMERA)!= PackageManager.PERMISSION_GRANTED){
            ActivityCompat.requestPermissions(faceRecognition.this, new String[]{
                    Manifest.permission.CAMERA
            }, 100);
        }
    }

    public void launchCameraIntent(){
        String fileName = "photo";
        File storageDirectory = getExternalFilesDir(Environment.DIRECTORY_PICTURES);
        try{
            File imageFile = File.createTempFile(fileName, ".jpg", storageDirectory);
            viewModel.currentPhotoPath = imageFile.getAbsolutePath();
            Uri imageUri = FileProvider.getUriForFile(faceRecognition.this, "com.example.thegoldenratio.fileprovider", imageFile);
            Intent intent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
            intent.putExtra(MediaStore.EXTRA_OUTPUT, imageUri);
            startActivityForResult(intent, 1);
        }catch (IOException e){
            e.printStackTrace();
        }
    }

    public void changeRectangle(int position){
        String text;
        TextView faceRecTextView = findViewById(R.id.faceRecTextView);

        final ImageView myImageView = binding.imageView;
        final Bitmap myBitmap = viewModel.mSelectedImage;
        final Paint myRectPaint;
        myRectPaint = new Paint();
        myRectPaint.setColor(Color.GREEN);
        myRectPaint.setStyle(Paint.Style.STROKE);
        myRectPaint.setStrokeWidth(STROKE_WIDTH);
        final float scale = getResources().getDisplayMetrics().density;

        final Bitmap tempBitmap = Bitmap.createBitmap(myBitmap.getWidth(), myBitmap.getHeight(), Bitmap.Config.RGB_565);
        final Canvas tempCanvas = new Canvas(tempBitmap);

        tempCanvas.drawBitmap(myBitmap, 0, 0, null);

        switch(position) {
            //for both eye and nose golden rectangle
            case(1):
                faceRecTextView.setText(viewModel.typeOfGoldenRatioText[0]);
                tempCanvas.drawRoundRect(new RectF((int) viewModel.faceCoordinates.noseTipX, (int) viewModel.faceCoordinates.eyeRightOuterY, (int) viewModel.faceCoordinates.eyeRightOuterX, (int) viewModel.faceCoordinates.noseTipY), 2, 2, myRectPaint);
                tempCanvas.drawLine((int) viewModel.faceCoordinates.eyeRightInnerX, (int) viewModel.faceCoordinates.eyeRightOuterY, (int) viewModel.faceCoordinates.eyeRightInnerX, (int) viewModel.faceCoordinates.noseTipY, myRectPaint);
                break;
            case(2):
                //for pupils and chin rectangle
                faceRecTextView.setText(viewModel.typeOfGoldenRatioText[1]);
                myRectPaint.setColor(Color.CYAN);
                double distBetweenPupils = viewModel.faceCoordinates.pupilRightX - viewModel.faceCoordinates.pupilLeftX;
                int rectHeight = (int)((distBetweenPupils)*(1.618));
                double smallRectHeight = distBetweenPupils/1.618;
                double smallRectY = viewModel.faceCoordinates.pupilLeftY + (rectHeight - smallRectHeight);
                tempCanvas.drawRoundRect(new RectF((int) viewModel.faceCoordinates.pupilLeftX, (int) viewModel.faceCoordinates.pupilLeftY, (int) viewModel.faceCoordinates.pupilRightX, (int) viewModel.faceCoordinates.pupilLeftY +rectHeight), 2, 2, myRectPaint);
                tempCanvas.drawLine((int) viewModel.faceCoordinates.pupilLeftX, (int)smallRectY, (int) viewModel.faceCoordinates.pupilRightX,  (int)smallRectY, myRectPaint);
                break;

            case(3):
                //for lips rect
                faceRecTextView.setText(viewModel.typeOfGoldenRatioText[2]);
                double lipsRectBreadth = viewModel.faceCoordinates.mouthRightX - viewModel.faceCoordinates.mouthLeftX;
                double lipsRectHeight = lipsRectBreadth/1.618;
                double lipsRectY = viewModel.faceCoordinates.mouthLeftY - lipsRectHeight/2;
                double lipsRectX = viewModel.faceCoordinates.mouthLeftX;
                double lipsRectXdas = viewModel.faceCoordinates.mouthRightX;
                double lipsRectYdas = lipsRectY+lipsRectHeight;
                double smallRectX = lipsRectX + (lipsRectBreadth/2.618);
                myRectPaint.setColor(Color.MAGENTA);
                tempCanvas.drawRoundRect(new RectF((int)lipsRectX, (int)lipsRectY, (int)lipsRectXdas, (int)lipsRectYdas), 2, 2, myRectPaint);
                tempCanvas.drawLine((float)smallRectX, (float)lipsRectY, (float)smallRectX, (float)lipsRectYdas, myRectPaint);
                break;
            case(4):
                //for pupilNoseRect
                faceRecTextView.setText(viewModel.typeOfGoldenRatioText[3]);
                double pnX = viewModel.faceCoordinates.pupilLeftX;
                double pnY = viewModel.faceCoordinates.pupilLeftY;
                double pnRectBreadth = viewModel.faceCoordinates.pupilRightX - viewModel.faceCoordinates.pupilLeftX;
                double pnRectHeight = pnRectBreadth/1.618;
                double pnXdas = viewModel.faceCoordinates.pupilRightX;
                double pnYdas = pnY+pnRectHeight;
                double smallpnRectY = pnRectHeight/2.618;
                myRectPaint.setColor(Color.BLUE);
                tempCanvas.drawRoundRect(new RectF((int)pnX, (int)pnY, (int)pnXdas, (int)pnYdas), 2, 2, myRectPaint);
                tempCanvas.drawLine((float)pnX, (float)(pnYdas-smallpnRectY), (float)pnXdas, (float)(pnYdas-smallpnRectY), myRectPaint);
                break;
            case(5):
                //for center-inner-width
                faceRecTextView.setText(viewModel.typeOfGoldenRatioText[4]);
                double ciwSmallRectX = viewModel.faceCoordinates.eyeLeftOuterX;
                double ciwSmallRectY = viewModel.faceCoordinates.eyeLeftOuterY;
                double ciwSmallRectDista = viewModel.faceCoordinates.noseTipX - viewModel.faceCoordinates.eyeLeftOuterX;
                double ciwSmallRectDistb = ciwSmallRectDista/1.618;
                double ciwRectX = viewModel.faceCoordinates.eyeLeftOuterX - ciwSmallRectDistb;
                double ciwRectY = viewModel.faceCoordinates.eyeLeftOuterY;
                double ciwRectBreadth = viewModel.faceCoordinates.noseTipX - ciwRectX;
                double ciwRectHeight = ciwRectBreadth * 1.618;
                double ciwRectXdas = ciwRectX + ciwRectBreadth;
                double ciwRectYdas = ciwRectY + ciwRectHeight;
                myRectPaint.setColor(Color.RED);
                tempCanvas.drawRoundRect(new RectF((int)ciwRectX, (int)ciwRectY, (int)ciwRectXdas, (int)ciwRectYdas), 2, 2, myRectPaint);
                tempCanvas.drawLine((float)ciwSmallRectX, (float)ciwSmallRectY, (float)ciwSmallRectX, (float)ciwSmallRectY + (float)ciwRectHeight, myRectPaint);
                break;
        }
        myImageView.setImageDrawable(new BitmapDrawable(getResources(), tempBitmap));

    }

}

File: app/src/main/java/com/example/thegoldenratio/facerecognition/faceRecognition.kt
package com.example.thegoldenratio.facerecognition;

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.*
import android.graphics.drawable.BitmapDrawable
import android.os.Bundle
import android.os.Environment
import android.provider.MediaStore
import android.util.Log
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.core.content.FileProvider
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import com.example.thegoldenratio.R
import com.example.thegoldenratio.databinding.ActivityFaceRecognitionBinding
import com.google.common.io.ByteStreams
import com.squareup.okhttp.MediaType
import com.squareup.okhttp.RequestBody
import retrofit2.HttpException
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.IOException

class faceRecognition: AppCompatActivity() {
    private lateinit var viewModel: FaceRecViewModel
    private lateinit var binding: ActivityFaceRecognitionBinding
    private var index = 1
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        permissions
        binding = DataBindingUtil.setContentView(this, R.layout.activity_face_recognition)
        viewModel = ViewModelProvider(this).get(FaceRecViewModel::class.java)
        binding.faceRecCameraFloatButton.setOnClickListener { launchCameraIntent() }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, intent: Intent?) {
        super.onActivityResult(requestCode, resultCode, intent)
        if (resultCode == RESULT_OK && requestCode == 1) {
            val bitmap = BitmapFactory.decodeFile(viewModel.currentPhotoPath)
            if (bitmap != null) {
                binding.imageView.visibility = View.VISIBLE
                viewModel.mSelectedImage = Bitmap.createScaledBitmap(
                    bitmap,
                    binding.imageView.width,
                    binding.imageView.height,
                    true
                )
                binding.imageView.setImageBitmap(viewModel.mSelectedImage)
                viewModel.analyzeImage()
                /*lifecycleScope.launchWhenCreated {
                    val response = try {
                        val outputStream = ByteArrayOutputStream()
                        val myBitmap: Bitmap = viewModel.mSelectedImage
                        myBitmap.compress(Bitmap.CompressFormat.JPEG, 100, outputStream)
                        val inputStream = ByteArrayInputStream(outputStream.toByteArray())

                        val localImageBytes: ByteArray = ByteStreams.toByteArray(inputStream)
                        val mediaType = MediaType.parse("application/octet-stream")
                        val body = RequestBody.create(
                            mediaType,
                            localImageBytes
                        )
                        RetrofitInstance.api.getFaceInfo(
                        body)
                    }catch (e: IOException) {
                        Log.e("asdf", e.toString())
                        return@launchWhenCreated
                    }catch (e: HttpException) {
                        Log.e("asdf", e.toString())
                        return@launchWhenCreated
                    }*/
                    /*    Log.i("asdf", response.toString())

                }*/
                try {
                    viewModel.analyzeJson(viewModel.resultString)
                    binding.forwardArrow.setOnClickListener {
                        index++
                        if (index > 5) index = 1
                        changeRectangle(index)
                    }
                    binding.backWardArrow.setOnClickListener {
                        index--
                        if (index < 1) index = 5
                        changeRectangle(index)
                    }
                } catch (e: Exception) {
                    Log.i("result", e.toString())
                }
            }
        }
    }

    private val permissions: Unit
        get() {
            if (ContextCompat.checkSelfPermission(
                    this@faceRecognition,
                    Manifest.permission.CAMERA
                ) != PackageManager.PERMISSION_GRANTED
            ) {
                ActivityCompat.requestPermissions(
                    this@faceRecognition, arrayOf(
                        Manifest.permission.CAMERA
                    ), 100
                )
            }
        }

    private fun launchCameraIntent() {
        val fileName = "photo"
        val storageDirectory = getExternalFilesDir(Environment.DIRECTORY_PICTURES)
        try {
            val imageFile = File.createTempFile(fileName, ".jpg", storageDirectory)
            viewModel.currentPhotoPath = imageFile.absolutePath
            val imageUri = FileProvider.getUriForFile(
                this@faceRecognition,
                "com.example.thegoldenratio.fileprovider",
                imageFile
            )
            val intent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
            intent.putExtra(MediaStore.EXTRA_OUTPUT, imageUri)
            startActivityForResult(intent, 1)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    private fun changeRectangle(position: Int) {
        var text: String
        val faceRecTextView = findViewById<TextView>(R.id.faceRecTextView)
        val myImageView = binding.imageView
        val myBitmap = viewModel.mSelectedImage
        val myRectPaint: Paint = Paint()
        myRectPaint.color = Color.GREEN
        myRectPaint.style = Paint.Style.STROKE
        myRectPaint.strokeWidth = STROKE_WIDTH
        val scale = resources.displayMetrics.density
        val tempBitmap = Bitmap.createBitmap(myBitmap.width, myBitmap.height, Bitmap.Config.RGB_565)
        val tempCanvas = Canvas(tempBitmap)
        tempCanvas.drawBitmap(myBitmap, 0f, 0f, null)
        when (position) {
            1 -> {
                faceRecTextView.text = viewModel.typeOfGoldenRatioText[0]
                tempCanvas.drawRoundRect(
                    RectF(
                        viewModel.faceCoordinates.noseTipX.toFloat(),
                        viewModel.faceCoordinates.eyeRightOuterY.toFloat(),
                        viewModel.faceCoordinates.eyeRightOuterX.toFloat(),
                        viewModel.faceCoordinates.noseTipY.toFloat()
                    ), 2f, 2f, myRectPaint
                )
                tempCanvas.drawLine(
                    (viewModel.faceCoordinates.eyeRightInnerX).toFloat(),
                    (viewModel.faceCoordinates.eyeRightOuterY).toFloat(),
                    (viewModel.faceCoordinates.eyeRightInnerX).toFloat(),
                    (viewModel.faceCoordinates.noseTipY).toFloat(),
                    myRectPaint
                )
            }
            2 -> {
                //for pupils and chin rectangle
                faceRecTextView.text = viewModel.typeOfGoldenRatioText[1]
                myRectPaint.color = Color.CYAN
                val distBetweenPupils =
                    viewModel.faceCoordinates.pupilRightX - viewModel.faceCoordinates.pupilLeftX
                val rectHeight = (distBetweenPupils * 1.618).toInt()
                val smallRectHeight = distBetweenPupils / 1.618
                val smallRectY =
                    viewModel.faceCoordinates.pupilLeftY + (rectHeight - smallRectHeight)
                tempCanvas.drawRoundRect(
                    RectF(
                        viewModel.faceCoordinates.pupilLeftX.toInt().toFloat(),
                        viewModel.faceCoordinates.pupilLeftY.toInt().toFloat(),
                        viewModel.faceCoordinates.pupilRightX.toInt().toFloat(),
                        (viewModel.faceCoordinates.pupilLeftY.toInt() + rectHeight).toFloat()
                    ), 2f, 2f, myRectPaint
                )
                tempCanvas.drawLine(
                    viewModel.faceCoordinates.pupilLeftX.toFloat(),
                    smallRectY.toFloat(),
                    viewModel.faceCoordinates.pupilRightX.toFloat(),
                    smallRectY.toFloat(),
                    myRectPaint
                )
            }
            3 -> {
                //for lips rect
                faceRecTextView.text = viewModel.typeOfGoldenRatioText[2]
                val lipsRectBreadth =
                    viewModel.faceCoordinates.mouthRightX - viewModel.faceCoordinates.mouthLeftX
                val lipsRectHeight = lipsRectBreadth / 1.618
                val lipsRectY = viewModel.faceCoordinates.mouthLeftY - lipsRectHeight / 2
                val lipsRectX = viewModel.faceCoordinates.mouthLeftX
                val lipsRectXdas = viewModel.faceCoordinates.mouthRightX
                val lipsRectYdas = lipsRectY + lipsRectHeight
                val smallRectX = lipsRectX + lipsRectBreadth / 2.618
                myRectPaint.color = Color.MAGENTA
                tempCanvas.drawRoundRect(
                    RectF(
                        lipsRectX.toFloat(),
                        lipsRectY.toFloat(),
                        lipsRectXdas.toFloat(),
                        lipsRectYdas.toFloat()
                    ), 2f, 2f, myRectPaint
                )
                tempCanvas.drawLine(
                    smallRectX.toFloat(),
                    lipsRectY.toFloat(),
                    smallRectX.toFloat(),
                    lipsRectYdas.toFloat(),
                    myRectPaint
                )
            }
            4 -> {
                //for pupilNoseRect
                faceRecTextView.text = viewModel.typeOfGoldenRatioText[3]
                val pnX = viewModel.faceCoordinates.pupilLeftX
                val pnY = viewModel.faceCoordinates.pupilLeftY
                val pnRectBreadth =
                    viewModel.faceCoordinates.pupilRightX - viewModel.faceCoordinates.pupilLeftX
                val pnRectHeight = pnRectBreadth / 1.618
                val pnXdas = viewModel.faceCoordinates.pupilRightX
                val pnYdas = pnY + pnRectHeight
                val smallpnRectY = pnRectHeight / 2.618
                myRectPaint.color = Color.BLUE
                tempCanvas.drawRoundRect(
                    RectF(
                        pnX.toFloat(),
                        pnY.toFloat(),
                        pnXdas.toFloat(),
                        pnYdas.toFloat()
                    ), 2f, 2f, myRectPaint
                )
                tempCanvas.drawLine(
                    pnX.toFloat(),
                    (pnYdas - smallpnRectY).toFloat(),
                    pnXdas.toFloat(),
                    (pnYdas - smallpnRectY).toFloat(),
                    myRectPaint
                )
            }
            5 -> {
                //for center-inner-width
                faceRecTextView.text = viewModel.typeOfGoldenRatioText[4]
                val ciwSmallRectX = viewModel.faceCoordinates.eyeLeftOuterX
                val ciwSmallRectY = viewModel.faceCoordinates.eyeLeftOuterY
                val ciwSmallRectDista =
                    viewModel.faceCoordinates.noseTipX - viewModel.faceCoordinates.eyeLeftOuterX
                val ciwSmallRectDistb = ciwSmallRectDista / 1.618
                val ciwRectX = viewModel.faceCoordinates.eyeLeftOuterX - ciwSmallRectDistb
                val ciwRectY = viewModel.faceCoordinates.eyeLeftOuterY
                val ciwRectBreadth = viewModel.faceCoordinates.noseTipX - ciwRectX
                val ciwRectHeight = ciwRectBreadth * 1.618
                val ciwRectXdas = ciwRectX + ciwRectBreadth
                val ciwRectYdas = ciwRectY + ciwRectHeight
                myRectPaint.color = Color.RED
                tempCanvas.drawRoundRect(
                    RectF(
                        ciwRectX.toFloat(),
                        ciwRectY.toFloat(),
                        ciwRectXdas.toFloat(),
                        ciwRectYdas.toFloat()
                    ), 2f, 2f, myRectPaint
                )
                tempCanvas.drawLine(
                    ciwSmallRectX.toFloat(),
                    ciwSmallRectY.toFloat(),
                    ciwSmallRectX.toFloat(),
                    ciwSmallRectY.toFloat() + ciwRectHeight.toFloat(),
                    myRectPaint
                )
            }
        }
        myImageView.setImageDrawable(BitmapDrawable(resources, tempBitmap))
    }

    companion object {
        private const val STROKE_WIDTH = 6.0f
    }
}